### PR TITLE
fix: fixed bug with rollout when starting with 100 instead of 0

### PR DESCRIPTION
**Pull request by jsalaber**
# Changes

- fixed bug when user doesn't qualify for next stage returning 0, when start percentage is non 0

### CODE REVIEW

File: lib/shared/bucketing/src/bucketing.ts [hunk 1]
```diff
@@ -105,7 +105,7 @@ export const getCurrentRolloutPercentage = (
             : null)
 
     if (!currentStage) {
-        return 0
+        return start
```
Comment (ajwootto) [inline lib/shared/bucketing/src/bucketing.ts:108]: pretty sure this just introduces a bug where it puts everyone in the start percentage of the rollout before the start date is actually triggered.

**Review by ajwootto — CHANGES_REQUESTED** [review]

File: lib/shared/bucketing/__tests__/bucketing.test.ts [hunk 2]
```diff
@@ -1415,6 +1415,118 @@ describe('Rollout Logic', () => {
...
+                        ),
+                        percentage: 1,
+                    },
+                ],
+            }
+
+            // Before next stage - should be 50%
+            jest.useFakeTimers().setSystemTime(new Date())
+            expect(
+                doesUserPassRollout({ rollout, boundedHash: 0.49 }),
```
Comment (ajwootto) [inline lib/shared/bucketing/__tests__/bucketing.test.ts:1475]: based on my understanding of how stepped rollouts are supposed to work, this test is wrong

Comment (jsalaber) [reply lib/shared/bucketing/__tests__/bucketing.test.ts:1475]: what's wrong with it?

Comment (ajwootto) [reply lib/shared/bucketing/__tests__/bucketing.test.ts:1475]: see slack

File: sdk/nextjs/src/common/transformConfig.ts [hunk 3]
```diff
@@ -0,0 +1,40 @@
+import { ConfigBody } from '@devcycle/types'
+
+/**
+ * Transform the config to ensure all dates are valid Date objects
+ * @param configData
+ * @returns
+ */
+export const transformConfig = (configData: ConfigBody): ConfigBody => {
```
Comment (ajwootto) [inline sdk/nextjs/src/common/transformConfig.ts:8]: why not use class-transformer to use the transformations already defined?

File: sdk/nextjs/src/pages/bucketing.ts [hunk 4]
```diff
@@ -52,7 +53,7 @@ const bucketOrFetchConfig = async (
 
     return generateBucketedConfig({
         user,
-        config,
+        config: transformConfig(config),
```
Comment (ajwootto) [inline sdk/nextjs/src/pages/bucketing.ts:56]: you should fix this at the fetching layer since the response is being asserted to be a ConfigBody, which it isn't

File: sdk/nextjs/src/common/transformConfig.ts (hunk 3 above)
Comment (ajwootto) [inline sdk/nextjs/src/common/transformConfig.ts:8]: the input arg here is not a ConfigBody since it hasn't been transformed into one yet

File: lib/shared/bucketing/__tests__/bucketing.test.ts [hunk 5]
```diff
@@ -1415,6 +1415,124 @@ describe('Rollout Logic', () => {
...
+            for (let i = 0; i < 100; i++) {
+                expect(
+                    doesUserPassRollout({ rollout, boundedHash: i / 100 }),
+                ).toBeTruthy()
+            }
+
+            jest.useRealTimers()
+        })
+
+        it.only('should handle stepped rollout with 50% start and 100% later stage', () => {
```
Comment (ajwootto) [inline lib/shared/bucketing/__tests__/bucketing.test.ts:1458]: remove it.only

File: lib/shared/bucketing/__tests__/bucketing.test.ts [hunk 6]
```diff
@@ -1415,6 +1415,124 @@ describe('Rollout Logic', () => {
...
+            for (let i = 0; i < 100; i++) {
+                expect(
+                    doesUserPassRollout({ rollout, boundedHash: i / 100 }),
+                ).toBeTruthy()
+            }
+
+            jest.useRealTimers()
+        })
+
+        it('should handle stepped rollout with 100% start and 0% later stage', () => {
```
Comment (ajwootto) [inline lib/shared/bucketing/__tests__/bucketing.test.ts:1498]: can you add a test case for when the starting step is in the future, to make sure it's zero

**Review by ajwootto — APPROVED** [review]
